cpp/src/arrow/buffer-builder.h

```diff
@@ -347,6 +347,9 @@ class TypedBufferBuilder<bool> {
   }
 
   Status Finish(std::shared_ptr<Buffer>* out, bool shrink_to_fit = true) {
+    // set bytes_builder_.size_ == byte size of data
+    bytes_builder_.UnsafeAdvance(BitUtil::BytesForBits(bit_length_) -
+                                 bytes_builder_.length());
```

Comment:
Do these bytes need to be zeroed?

Reply:
`TypedBufferBuilder<bool>` doesn't utilize `bytes_builder_`'s append methods, which means the size of bytes_builder_ is unchanged.

I'm thinking it's not worthwhile to use `BufferBuilder` in this class, I'm going to refactor and instead just manage a buffer directly

Reply:
OK, please do that refactoring in a follow up PR and not this one =)